File: 31-60/jumpGame55.py

def canJump(nums) -> bool:
    goal = len(nums) -1
    print('goal', goal)

    # starting from the last index to the begining index
    for i in range(len(nums)-2, -1, -1): 
        print('i', i, 'nums[i]', nums[i])
        if i + nums[i] >= goal:
            goal = i
            print('goal111', goal)
            
    return True if goal == 0 else False


nums = [2,3,1,1,4]
nums1 = [3,2,1,0,4]
    
print(canJump(nums1))

File: jumpGame55.java
public class jumpGame55 {
    public static boolean jumpGame(int[] nums) {
        int goal = nums.length - 1;
        for (int i = nums.length - 2; i >= 0; i --) {
            if (i + nums [i] >= goal) goal = i;
        }
        if (goal == 0) return true;
        else return false;
    }

    public static void main(String[] args) {
        int[] nums = {2,3,1,1,4};
        System.out.println(jumpGame(nums));
    }
}
